File: src/algorithms/recursion/powerset/solution2.js
// O(n*2^n) time | O(n*2^n) space
function powerset(array, idx = null) {
  if (idx === null) {
  	idx = array.length - 1;
  };
  if (idx < 0) {
  	return [[]];
  };
  const ele = array[idx];
  const subsets = powerset(array, idx - 1);
  const length = subsets.length;
  for (let i = 0; i < length; i++) {
  	const currentSubset = subsets[i];
  	subsets.push(currentSubset.concat(ele));
  };
  return subsets;
};

exports.powerset = powerset;

File: src/algorithms/recursion/powerset/solution2.go
package main

// O(n*2^2) time | O(n*2^2) space
func Powerset(array []int) [][]int {
	subsets := [][]int{{}}
	for _, ele := range array {
		length := len(subsets)
		for i := 0; i < length; i++ {
			currentSubset := subsets[i]
			newsubset := append([]int{}, currentSubset...)
			newsubset = append(newsubset, ele)
			subsets = append(subsets, newsubset)
		}
	}
	return subsets
}